생성형 AI 도구 링크 :
https://claude.ai/share/f28fb6b9-354f-4b1c-a14a-f17aba815321 https://claude.ai/share/f013d490-a5c3-40d1-a009-42de61c93b09 (길이 제한이 있어 나눠서 올립니다.)

Q1

(1)코드.

In [5]:
import csv, json, logging

logging.basicConfig(level=logging.INFO)

def make_report(csv_path: str, json_path: str) -> int:
    results: list = []
    success_count: int = 0

    try:
        with open(csv_path, "r" , encoding="utf-8") as f:
            reader = csv.DictReader(f)
            for row in reader:
                name: str = row["이름"]
                sid: str = row["학번"]
                mid_str: str = row["중간"]
                fin_str: str = row["기말"]
                hw_str: str = row["과제"]

                if mid_str == "" or fin_str == "" or hw_str == "":
                    mid = None if mid_str == "" else int(mid_str)
                    fin = None if fin_str == "" else int(fin_str)
                    hw = None if hw_str == "" else int(hw_str)
                    avg = None
                    grade = None
                    logging.info(f"{name}: 평균 None, 등급 None (결측값)")
                else:
                    mid: int = int(mid_str)
                    fin: int = int(fin_str)
                    hw: int = int(hw_str)
                    avg: float = mid * 0.3 + fin * 0.5 + hw * 0.2

                    if avg >= 90:
                        grade: str = "A"
                    elif avg >=80:
                        grade: str = "B"
                    elif avg >=70:
                        grade: str = "C"
                    else:
                        grade: str = "F"

                    logging.info(f"{name}: 평균 {avg}, 등급 {grade}")
                    success_count +=1

                results.append({
                    "이름": name,
                    "학번": sid,
                    "점수": {"중간": mid, "기말": fin, "과제": hw},
                    "평균": avg,
                    "등급": grade,
                }) 
    except FileNotFoundError:
        logging.warning(f"CSV 파일이 없습니다: {csv_path}")
        return 0
    except UnicodeDecodeError:
        logging.error(f"CSV 파일의 인코딩이 잘못되었습니다: {csv_path}")
        return 0
    
    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(results, f, ensure_ascii=False, indent=2)

    return success_count

n = make_report("scores.csv", "report.json")
print(f"성공 처리 학생 수: {n}")

with open("report.json", "r", encoding="utf-8") as f:
    print(f.read())

INFO:root:김언어: 평균 89.5, 등급 B
INFO:root:이국문: 평균 84.4, 등급 B
INFO:root:박영문: 평균 93.5, 등급 A
INFO:root:최역사: 평균 None, 등급 None (결측값)


성공 처리 학생 수: 3
[
  {
    "이름": "김언어",
    "학번": "2026-10000",
    "점수": {
      "중간": 85,
      "기말": 92,
      "과제": 90
    },
    "평균": 89.5,
    "등급": "B"
  },
  {
    "이름": "이국문",
    "학번": "2026-12345",
    "점수": {
      "중간": 78,
      "기말": 88,
      "과제": 85
    },
    "평균": 84.4,
    "등급": "B"
  },
  {
    "이름": "박영문",
    "학번": "2026-13579",
    "점수": {
      "중간": 95,
      "기말": 90,
      "과제": 100
    },
    "평균": 93.5,
    "등급": "A"
  },
  {
    "이름": "최역사",
    "학번": "2025-11111",
    "점수": {
      "중간": null,
      "기말": 82,
      "과제": 88
    },
    "평균": null,
    "등급": null
  }
]


(2)실행결과. INFO:root:김언어: 평균 89.5, 등급 B
INFO:root:이국문: 평균 84.4, 등급 B
INFO:root:박영문: 평균 93.5, 등급 A
INFO:root:최역사: 평균 None, 등급 None (결측값)
성공 처리 학생 수: 3
[
  {
    "이름": "김언어",
    "학번": "2026-10000",
    "점수": {
      "중간": 85,
      "기말": 92,
      "과제": 90
    },
    "평균": 89.5,
    "등급": "B"
  },
  {
    "이름": "이국문",
    "학번": "2026-12345",
    "점수": {
      "중간": 78,
      "기말": 88,
      "과제": 85
    },
    "평균": 84.4,
    "등급": "B"
  },
  {
    "이름": "박영문",
    "학번": "2026-13579",
    "점수": {
      "중간": 95,
      "기말": 90,
      "과제": 100
    },
    "평균": 93.5,
    "등급": "A"
  },
  {
    "이름": "최역사",
    "학번": "2025-11111",
    "점수": {
      "중간": null,
      "기말": 82,
      "과제": 88
    },
    "평균": null,
    "등급": null
  }
]

(3)설명. csv.DictReader로 읽을 때 빈 칸을 빈 문자열 ""로 읽어오는 점을 이용해, 세 점수 중 하나라도 ""이면 결측값으로 보고 그 학생의 평균과 등급을 None으로 두었다. 채워진 점수만 int()로 변환해 보존했고, json.dump에서 파이썬의 None은 자동으로 JSON의 null로 변환된다. 인코딩은 문제에서 score.csv가 "UTF-8, BOM 없음"이라고 명시되어 있어 한글 텍스트의 표준인 encoding="utf-8"을 지정했고, JSON 출력에는 한글이 깨지지 않도록 ensure_ascii=False, 가독성을 위해 indent=2를 함께 사용했다. 

Q2

(1)코드.

In [1]:
class InvalidJamoError(ValueError):
    """한글 자모가 아닌 문자가 입력되었을 때 발생하는 예외."""

def classify_jamo(c: str) -> str:
    if not isinstance(c, str):
        raise TypeError(f"문자열이 아닙니다: {type(c).__name__}")
    if len(c) != 1:
        raise ValueError(f"길이가 1이 아닙니다: {c!r}")
    code: int = ord(c)
    if 0x3131 <= code <= 0x314E:
        return "자음"
    elif 0x314F <= code <= 0x3163:
        return "모음"
    else:
        raise InvalidJamoError(f"한글 자모가 아닙니다: {c!r}")

inputs = ["ㄱ", "ㅏ", "ㄲ", "가", "AB", 5, "ㅎ","ㅣ",""]

for item in inputs:
    try:
        result: str = classify_jamo(item)
        print(f"{item!r}: {result}")
    except TypeError as e:
        print(f"[TypeError] {e}")
    except InvalidJamoError as e:
        print(f"[InvalidJamoError] {e}")
    except ValueError as e:
        print(f"[ValueError] {e}")

'ㄱ': 자음
'ㅏ': 모음
'ㄲ': 자음
[InvalidJamoError] 한글 자모가 아닙니다: '가'
[ValueError] 길이가 1이 아닙니다: 'AB'
[TypeError] 문자열이 아닙니다: int
'ㅎ': 자음
'ㅣ': 모음
[ValueError] 길이가 1이 아닙니다: ''


(2)실행결과. 'ㄱ': 자음
'ㅏ': 모음
'ㄲ': 자음
[InvalidJamoError] 한글 자모가 아닙니다: '가'
[ValueError] 길이가 1이 아닙니다: 'AB'
[TypeError] 문자열이 아닙니다: int
'ㅎ': 자음
'ㅣ': 모음
[ValueError] 길이가 1이 아닙니다: ''

(3)설명. InvalidJamoError는 "값이 도메인에 맞지 않다"는 의미이므로, 같은 의미를 갖는 표준 예외 ValueError의 자식으로 정의하는 것이 자연스럽다. 이렇게 하면 호출 측에서 InvalidJamoError로 좁게 잡거나 ValueError로 넓게 잡을 수 있어 구체적인 예외 종류를 명시할 수 있으며, 부모 예외로 자식까지 한꺼번에 잡을 수 있다는 계층 구조의 이점을 함께 누릴 수 있다.